ttnn/cpp/ttnn/operations/moreh/moreh_linear_backward/moreh_linear_backward.cpp

```diff
@@ -0,0 +1,186 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "moreh_linear_backward.hpp"
+
+#include "ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/moreh_helper_functions.hpp"
+#include "ttnn/cpp/ttnn/operations/moreh/moreh_matmul/moreh_matmul.hpp"
+#include "ttnn/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum.hpp"
+
+namespace ttnn::operations::moreh::moreh_linear_backward {
+
+std::tuple<bool, bool, bool> MorehLinearBackward::get_required_outputs(const std::vector<bool>& are_required_outputs) {
+    if (are_required_outputs.size() != 3) {
+        TT_ASSERT(are_required_outputs.size() == 3, "are_required_outputs size must be 3");
+    }
+
+    return {are_required_outputs[0], are_required_outputs[1], are_required_outputs[2]};
+}
+
+void get_tensor_dim(std::vector<uint32_t>& dim, const tt::tt_metal::LegacyShape& shape) {
+    const auto rank = shape.rank();
+    for (auto i = 0; i < rank; ++i) {
+        auto idx = rank - 1 - i;
+
+        // last 2-dim
+        if (idx == rank - 1 || idx == rank - 2) {
+            dim[i] = shape[idx] / tt::constants::TILE_HEIGHT;
+        } else {
+            dim[i] = shape[idx];
+        }
+    }
+
+    log_debug(tt::LogOp, "rank {}", rank);
+    for (auto i = 0; i < tt::tt_metal::MAX_NUM_DIMENSIONS; ++i) {
+        log_debug(tt::LogOp, "dim[{}] = {}", i, dim[i]);
+    }
+}
+
+inline void moreh_linear_backward_validate(
+    const Tensor& output_grad,
+    const Tensor& input,
+    const Tensor& weight,
+    const std::optional<const Tensor>& input_grad,
+    const std::optional<const Tensor>& weight_grad,
+    const std::optional<const Tensor>& bias_grad) {
+    if (input_grad.has_value()) {
+        const auto& input_grad_tensor = input_grad.value();
+        TT_ASSERT(
+            tt::operations::primary::is_same_shape(input, input_grad_tensor), "both tensors should be the same shape");
+    }
+
+    if (weight_grad.has_value()) {
+        const auto& weight_grad_tensor = weight_grad.value();
+        TT_ASSERT(
```

Comment:
```suggestion
        TT_FATAL(
```

Reply:
Resolved in https://github.com/tenstorrent/tt-metal/pull/12812/commits/8dec1ff696bd9948f9366cbee8ec5733a58f2081